docs/architecture/adr-100-data-companion-push-api.md

```diff
@@ -0,0 +1,441 @@
+
+The following simple diagram shows the proposed relationships between
+Tendermint, a socket-based ABCI application, and the proposed data companion
+service.
+
+```
+     +----------+      +------------+      +----------------+
+     | ABCI App | <--- | Tendermint | ---> | Data Companion |
+     +----------+      +------------+      +----------------+
+
+```
+
+As can be seen in this diagram, Tendermint connects out to both the ABCI
+application and data companion service based on the Tendermint node's
+configuration.
+
+The fact that Tendermint connects out to the companion service instead of the
+other way around provides a natural constraint on the number of consumers of the
+API. This also implies that the data companion is a trusted external service,
+just like an ABCI application, as it would need to be configured within the
+Tendermint configuration file.
+
+### gRPC API
+
+The following gRPC API is proposed:
+
+```protobuf
+import "tendermint/abci/types.proto";
+import "tendermint/types/block.proto";
+
+// DataCompanionService allows Tendermint to publish certain data generated by
+// the consensus engine to a single external consumer with specific reliability
+// guarantees.
+//
+// Note that implementers of this service must take into account the possibility
+// that Tendermint may re-send data that was previously sent. Therefore
+// the service should simply ignore previously seen data instead of responding
+// with errors to ensure correct functioning of the node, as responding with an
+// error is taken as a signal to Tendermint that it must crash and await
+// operator intervention.
+service DataCompanionService {
+    // BlockCommitted is called after a block has been committed. This method is
+    // also called on Tendermint startup to ensure that the service received the
+    // last committed block, in case there was previously a transport failure.
+    //
+    // If an error is returned, Tendermint will crash.
+    rpc BlockCommitted(BlockCommittedRequest) returns (BlockCommittedResponse) {}
+}
+
+// BlockCommittedRequest contains at least the data for the block that was just
+// committed. If enabled, it also contains the ABCI FinalizeBlock response data
+// related to the block in this request.
+message BlockCommittedRequest {
+    // The block data for the block that was just committed.
+    tendermint.types.Block                         block = 1;
+    // The FinalizeBlockResponse related to the block in this request. This
+    // field is optional, depending on the Tendermint node's configuration.
+    optional tendermint.abci.FinalizeBlockResponse finalize_block_response = 2;
+}
+
+// BlockCommittedResponse is either empty upon succes, or returns one or more
+// errors. Note that returning any errors here will cause Tendermint to crash.
+message BlockCommittedResponse {
+    // Zero or more errors occurred during the companion's processing of the
+    // request.
+    repeated Error errors = 1;
```

Comment:
Not sure about this, but I'd give it a thought about dropping "repeated" here, since we are crashing.